### PR TITLE
Add new tools package, including subsystem wizard

### PR DESCRIPTION
**Pull request by davidkline-ms**
This change adds the com.microsoft.mrtk.tools package (equivalent to MRTK2's tools package). In this submission, the package contains only the new Subsystem Wizard.

The subsystem wizard replaces the MRTK2 Extension Service Creator and makes it easy to create the scaffolding for a new Unity XR Subsystem.

image

image

image

image

### CODE REVIEW

File: com.microsoft.mrtk.tools/package.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "com.microsoft.mrtk.tools",
+  "version": "3.0.0-development",
```
Comment (Zee2) [inline com.microsoft.mrtk.tools/package.json:3]: This should be `1.0.0-development` if the package was not present before in any prior release of MRTK, right?

File: com.microsoft.mrtk.tools/SubsystemWizard/Templates/SubsystemInterfaceTemplate.txt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// TODO: [Optional] Add additional copyright statement(s).
+
+namespace %NAMESPACE%
+{
+    /// <summary>
+    /// Interface defining the functionality of the #SUBSYSTEMBASECLASSNAME#.
```
Comment (Zee2) [inline com.microsoft.mrtk.tools/SubsystemWizard/Templates/SubsystemInterfaceTemplate.txt:8]: It seems like `#TEMPLATE#` and `%TEMPLATE%` are mixed and matched. Are there supposed to be two different template symbols/formats?

File: com.microsoft.mrtk.tools/SubsystemWizard/Templates/SubsystemConfigTemplate.txt [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (Zee2) [inline com.microsoft.mrtk.tools/SubsystemWizard/Templates/SubsystemConfigTemplate.txt:1]: Do we want to insert Copyright MSFT on new templates? Shouldn't it technically be `Copyright (c) %COMPANYNAME%` or something like that?

Do we even want to automatically apply copyright notices to people's code, even if it's their organization name? Sort of gives a strange impression ("you write an extension to MRTK? COPYRIGHT IT NOW!")

File: com.microsoft.mrtk.tools/SubsystemWizard/Templates/SubsystemBaseClassTemplate.txt [hunk 4]
```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (Zee2) [inline com.microsoft.mrtk.tools/SubsystemWizard/Templates/SubsystemBaseClassTemplate.txt:1]: Etc on all the copyright notices.

File: UnityProjects/MRTKDevTemplate/Packages/packages-lock.json [hunk 5]
```diff
@@ -179,7 +187,7 @@
       "url": "https://packages.unity.com"
     },
     "com.unity.collab-proxy": {
-      "version": "1.15.18",
+      "version": "1.17.0",
```
Comment (Zee2) [inline UnityProjects/MRTKDevTemplate/Packages/packages-lock.json:190]: Did you mean to check these in? This usually happens when you're opening the project in a newer version of Unity

File: UnityProjects/MRTKDevTemplate/Packages/manifest.json [hunk 6]
```diff
@@ -14,13 +14,14 @@
     "com.microsoft.mrtk.input": "file:../../../com.microsoft.mrtk.input",
     "com.microsoft.mrtk.spatialmanipulation": "file:../../../com.microsoft.mrtk.spatialmanipulation",
     "com.microsoft.mrtk.standardassets": "file:../../../com.microsoft.mrtk.standardassets",
+    "com.microsoft.mrtk.tools": "file:../../../com.microsoft.mrtk.tools",
     "com.microsoft.mrtk.uxcomponents": "file:../../../com.microsoft.mrtk.uxcomponents",
     "com.microsoft.mrtk.uxcore": "file:../../../com.microsoft.mrtk.uxcore",
     "com.microsoft.mrtk.windowsspeech": "file:../../../com.microsoft.mrtk.windowsspeech",
     "com.microsoft.spatialaudio.spatializer.unity": "file:../../../ExternalDependencies/com.microsoft.spatialaudio.spatializer.unity-1.0.246.tgz",
-    "com.unity.collab-proxy": "1.15.18",
+    "com.unity.collab-proxy": "1.17.0",
```
Comment (Zee2) [inline UnityProjects/MRTKDevTemplate/Packages/manifest.json:22]: Etc, did you mean to bump these?

**Review by Zee2 — CHANGES_REQUESTED** [review]
Only blocking one is the 3.0.0 vs 1.0.0 versioning thing.

File: com.microsoft.mrtk.tools/package.json (hunk 1 above)
Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/package.json:3]: we had a tools package in MRTK2, so i started this one at 3.0.0

File: com.microsoft.mrtk.tools/SubsystemWizard/Templates/SubsystemConfigTemplate.txt (hunk 3 above)
Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/Templates/SubsystemConfigTemplate.txt:1]: i debated on that one.... i think i'll remove them :)

File: UnityProjects/MRTKDevTemplate/Packages/packages-lock.json (hunk 5 above)
Comment (davidkline-ms) [reply UnityProjects/MRTKDevTemplate/Packages/packages-lock.json:190]: i think these happened when i used the latest 2020 lts

Comment (davidkline-ms) [reply UnityProjects/MRTKDevTemplate/Packages/packages-lock.json:190]: i can revert

File: com.microsoft.mrtk.tools/SubsystemWizard/Templates/SubsystemInterfaceTemplate.txt (hunk 2 above)
Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/Templates/SubsystemInterfaceTemplate.txt:6]: there should be only one.... my error

Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/Templates/SubsystemInterfaceTemplate.txt]: fixed

File: com.microsoft.mrtk.tools/package.json (hunk 1 above)
Comment (Zee2) [reply com.microsoft.mrtk.tools/package.json:3]: Had no idea! LOL

File: UnityProjects/MRTKDevTemplate/Packages/manifest.json [hunk 7]
```diff
@@ -14,13 +14,14 @@
     "com.microsoft.mrtk.input": "file:../../../com.microsoft.mrtk.input",
     "com.microsoft.mrtk.spatialmanipulation": "file:../../../com.microsoft.mrtk.spatialmanipulation",
     "com.microsoft.mrtk.standardassets": "file:../../../com.microsoft.mrtk.standardassets",
+    "com.microsoft.mrtk.tools": "file:../../../com.microsoft.mrtk.tools",
```
Comment (Zee2) [inline UnityProjects/MRTKDevTemplate/Packages/manifest.json:17]: Do we need to include the tools package in the template project?

File: UnityProjects/MRTKDevTemplate/ProjectSettings/ProjectVersion.txt [hunk 8]
```diff
@@ -1,2 +1,2 @@
-m_EditorVersion: 2020.3.36f1
```
Comment (Zee2) [inline UnityProjects/MRTKDevTemplate/ProjectSettings/ProjectVersion.txt:1]: Are we good bumping this?

File: com.microsoft.mrtk.core/Editor/Inspectors/Utilities/MixedRealityInspectorUtility.cs [hunk 9]
```diff
@@ -55,17 +55,14 @@ public static Color DefaultBackgroundColor
...
-        private static readonly GUIStyle ProductNameStyle = new GUIStyle(EditorStyles.boldLabel)
+        public static readonly GUIStyle ProductNameStyle = new GUIStyle(EditorStyles.boldLabel)
         {
             fontSize = 26,
             alignment = TextAnchor.UpperCenter,
             fixedHeight = 32
         };
 
-        internal static bool IsMixedRealityToolkitLogoAssetPresent()
+        public static bool IsMixedRealityToolkitLogoAssetPresent()
```
Comment (keveleigh) [inline com.microsoft.mrtk.core/Editor/Inspectors/Utilities/MixedRealityInspectorUtility.cs:65]: does this need to be public or could we use `InternalsVisibleTo`? It feels like a really specific method that others shouldn't need to worry about

File: com.microsoft.mrtk.tools/package.json [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  "description": "A collection of tools, utilities and wizards that can be used to create and analyze components for use with MRTK3.",
```
Comment (keveleigh) [inline com.microsoft.mrtk.tools/package.json:4]: nit: Oxford comma 🙃

```suggestion
  "description": "A collection of tools, utilities, and wizards that can be used to create and analyze components for use with MRTK3.",
```

File: UnityProjects/MRTKDevTemplate/Packages/manifest.json (hunk 7 above)
Comment (davidkline-ms) [reply UnityProjects/MRTKDevTemplate/Packages/manifest.json:17]: we do not. it was there for testing

File: UnityProjects/MRTKDevTemplate/ProjectSettings/ProjectVersion.txt (hunk 8 above)
Comment (davidkline-ms) [reply UnityProjects/MRTKDevTemplate/ProjectSettings/ProjectVersion.txt:1]: confirmed with CI, we should be fine

File: UnityProjects/MRTKDevTemplate/Packages/manifest.json (hunk 7 above)
Comment (keveleigh) [reply UnityProjects/MRTKDevTemplate/Packages/manifest.json:17]: It wouldn't hurt to keep it around imo, so we can use the tools ourselves if want!

File: com.microsoft.mrtk.core/Editor/Inspectors/Utilities/MixedRealityInspectorUtility.cs (hunk 9 above)
Comment (davidkline-ms) [reply com.microsoft.mrtk.core/Editor/Inspectors/Utilities/MixedRealityInspectorUtility.cs:65]: i can make that change

File: UnityProjects/MRTKDevTemplate/Packages/manifest.json (hunk 7 above)
Comment (davidkline-ms) [reply UnityProjects/MRTKDevTemplate/Packages/manifest.json:17]: works for me. i will leave it there :)

File: com.microsoft.mrtk.core/Editor/Inspectors/Utilities/MixedRealityInspectorUtility.cs (hunk 9 above)
Comment (davidkline-ms) [reply com.microsoft.mrtk.core/Editor/Inspectors/Utilities/MixedRealityInspectorUtility.cs]: done.

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs [hunk 11]
```diff
@@ -0,0 +1,605 @@
...
+        // SubsystemWizard/Templates/SubsystemDerivedClassTemplate.txt
+        private const string DerivedClassTemplateGuid = "02b67503c64cf224ca2c04a04077c0d7";
+        // SubsystemWizard/Templates/SubsystemDescriptorTemplate.txt
+        private const string DescriptorTemplateGuid = "8e1afb29fbaf1c2419511d266f49b976";
+        // SubsystemWizard/Templates/SubsystemInterfaceTemplate.txt
+        private const string InterfaceTemplateGuid = "9bac30e514266984d947e360cfd17b05";
+
+        private const bool DefaultCreateConfiguration = false;
+        private static readonly string DefaultBaseSubsystemName = $"NewSubsystem";
+        private static readonly string DefaultOrganizationName = "Contoso";
```
Comment (keveleigh) [inline com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs:36]: Would we maybe want to default to https://docs.unity3d.com/ScriptReference/PlayerSettings-companyName.html?

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs [hunk 12]
```diff
@@ -0,0 +1,605 @@
...
+        /// </summary>
+        public string DisplayName => $"{OrganizationName} {BaseClassName}";
+
+        /// <summary>
+        /// Name of interface to create for new subystem.
+        /// </summary>
+        public string InterfaceName
+        {
+            get => $"I{BaseClassName}";
+        }
```
Comment (keveleigh) [inline com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs:115]: nit: mixing styles here (and a couple other places) 🙂 I personally prefer the top one!

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs [hunk 13]
```diff
@@ -0,0 +1,605 @@
...
+
+        /// <summary>
+        /// Validates that the specified name is a valid indentifier for C#.
+        /// </summary>
+        /// <param name="name">The name to validate/</param>
+        /// <returns>True if the name is a valid C# identifier, or false.</returns>
+        private bool ValidateName(string name)
+        {
+            // Verify that the name is valid within C#
+                                        return CSharpCodeProvider.CreateProvider("C#").IsValidIdentifier(name);
```
Comment (keveleigh) [inline com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs:581]: nit: weird indentation here

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs (hunk 12 above)
Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs:115]: will fix

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs (hunk 13 above)
Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs:581]: that IS odd.... i havent seen that one 😆

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs (hunk 11 above)
Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs:36]: ooh. i like that.

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs (hunk 12 above)
Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs]: fixed

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs (hunk 13 above)
Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs:572]: looks like something landed on my spacebar at some point... haha

Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs:572]: fixed

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs (hunk 11 above)
Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs]: done. thx for the idea!

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs [hunk 14]
```diff
@@ -0,0 +1,595 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Microsoft.CSharp;
```
[the remark this answers is not in the document]
Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs:4]: not sure. this is something i took from the MRTK2 extension service creator.

[the remark this answers is not in the document]
Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs:4]: i do not recall seeing any issues filed related to that tool and non-Windows editors

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs [hunk 15]
```diff
@@ -0,0 +1,595 @@
...
+
+        /// <summary>
+        /// Indicates if the user wishes to skip the creation of the subsystem
+        /// descriptor source code file.
+        /// </summary>
+        public bool DontCreateImplementationClass
+        {
+            get => dontCreateImplementationClass;
+            set => dontCreateImplementationClass = value;
+        }
```
Comment (keveleigh) [inline com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs:153]: super nit: could these be auto-properties?

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs [hunk 16]
```diff
@@ -0,0 +1,595 @@
...
+        /// Ensures that the necessary templates can be located and successfully loaded.
+        /// </summary>
+        /// <param name="errors">List to which any encountered errors will be added.</param>
+        /// <param name="descriptorTemplate">
+        /// <see cref="FileInfo"/> object representing the subsystem descriptor template file.
+        /// </param>
+        /// <param name="interfaceTemplate">
+        /// <see cref="FileInfo"/> object representing the subsystem interface template file.
+        /// </param>
+        /// <param name="classTemplate">
```
Comment (keveleigh) [inline com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs:349]: I don't think this param exists anymore (@MaxWang-MS, is there any way for the docs gen pipeline to treat something like this as an error?)

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemWizardWindow.cs [hunk 17]
```diff
@@ -0,0 +1,416 @@
...
+using UnityEngine;
+using UnityEditor;
+
+namespace Microsoft.MixedReality.Toolkit.Tools
+{
+    /// <summary>
+    /// Editor window to guide developers through the process of creating a
+    /// Unity XR Subsystem for use with MRTK3.
+    /// </summary>
+    public class SubsystemWizardWindow : EditorWindow
```
Comment (keveleigh) [inline com.microsoft.mrtk.tools/SubsystemWizard/SubsystemWizardWindow.cs:18]: Does this need to be public?

**Review by keveleigh — DISMISSED** [review]

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs (hunk 15 above)
Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs:153]: yep. just a habit when writing mrtk code 😄

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemWizardWindow.cs (hunk 17 above)
Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/SubsystemWizardWindow.cs:18]: nope

Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/SubsystemWizardWindow.cs]: now intrernal

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs (hunk 16 above)
Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs]: forgot to rename the param in the doc comment

File: com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs (hunk 15 above)
Comment (davidkline-ms) [reply com.microsoft.mrtk.tools/SubsystemWizard/SubsystemGenerator.cs]: updated

**Review by Zee2 — APPROVED** [review]

**Pull request merged**